lib/models/position.rb

```diff
@@ -1,21 +1,49 @@
 class Position < ActiveRecord::Base
 
-  def grid_id
-    # TODO
+  before_validation :compute_grid_id
+
+  # A computed grid_id must be present
+  validates :grid_id, presence: true
+
+  def travel_times
+    TravelTime.where(input_id: grid_id)
   end
 
-  def location
-    # TODO [y, x]
+  def self.random
+    order('RANDOM()')
   end
 
   def self.available(run)
-    # where not in run.placements
+    # where not in run.placements -- there's probably a prettier way
+    where("id not in (:placements)", placements: run.placements.pluck(:position_id))
   end
 
-  def self.within(time = 40.minutes, of: , via: :walking)
+  def within?(time = 40.minutes, of: , via: :walking)
     # TODO: Guard that given object has grid_id.
     # where lookup grid cells, travel time between
     # applicant and position is less than time given
     time = time.to_i
+    travel_times.where(travel_mode: via)
+                .where("time < #{time}")
+                .exists?(target_id: of.grid_id) if of.respond_to?(:grid_id)
   end
+
+  def self.within(time = 40.minutes, of: , via: :walking)
+    # TODO: Guard that given object has grid_id.
+    time = time.to_i
+    ids = of.travel_times
+            .where(travel_mode: via)
+            .where("time < #{time}")
+            .pluck(:target_id)
+            .uniq if of.respond_to?(:grid_id)
+
+    self.where(grid_id: ids)
```

Comment:
There's enough duplicated logic between `self.within` and `within?` that we could refactor a little.